### ISSUE TITLE
Reduce execution time when reading large datasets

### ISSUE THREAD

**Issue by alvmas** [nanograv/PINT#1002]
Hello,

I am a member of the high energy observatory CTA. I am trying to use PINT to calculate the pulsar phase associated with the time of arrival of each photon detected at Very High Energies. 

I can see that PINT works somehow slow for large datasets so it could become inefficient for long time observations and I was wondering if there could be an easiest way to make the process faster.

@rlopezcoto and I have taken a look at the code and we think that the problem might be that TOAs objects are created reading one by one the times in the toafile so one possible solution could be to create arrays of toas that could be read faster and work with all the timing corrections through arrays. Could this be a good option? Is this actually easy to implement in the current code?

Do you think there could be another way to enhance the time of execution for large datasets? 

Thanks in advance,
Álvaro Mas

**Comment by paulray** [comment]
Something like this is potentially possible. @kerrm has thought about it some for Fermi TOAs.

Reading large numbers is kind of slow, but if you are processing your TOAs more than once, the TOA pickling feature makes all subsequent reads of the same TOAs very fast.

One of the speed issues is having to create a `Time` object for each photon instead of one `Time` object that holds an array of photons. The reason the latter is not possible is that when a `Time` holds an array of times, all of them must have the same `location`.

**Comment by kerrm** [comment]
How many events are we talking about here?  I've scaled up to a few million with no problem with Fermi data.   The TOA instantiation is somewhat faster than it used to be.  But I agree it's still pretty slow compared to e.g. tempo2, and some development there would be great.

Like Paul mentioned, we're missing a piece of astropy functionality that would be needed to instantiate Time objects with arrays, so perhaps that's actually a better place to think about a pull request!

**Comment by alvmas** [comment]
Thanks for the response!  I can see that the problem is related with the creation of the Time object. I tried to work with sets of a few million data as well, each one takes approx 2 hours or more to create the TOAs object. @kerrm does it take you approximately such a large time?

**Comment by kerrm** [comment]
That sounds about right.  My pipeline parallelizes the computation, pickles
the TOAs from each segments, and then merges them, so it's somewhat faster
in real time.  As Paul mentioned, once those TOAs are pickled, though, it's
much faster (~1 to 2min for that many TOAs) to re-load them, so I'm not
really impacted by the slow initial time.

Anyway, it would still be nicer if it were faster.

**Comment by aarchiba** [comment]
There are certainly inefficiencies in PINT's work. But one of the slow steps is the timescale conversion to TDB and computation of the Earth position for each TOA, which is a somewhat expensive calculation regardless of how many Time objects are created. The TOA pickling function pickles them after this calculation, so it avoids having to redo this computation.

**Comment by rlopezcoto** [comment]
> One of the speed issues is having to create a `Time` object for each photon instead of one `Time` object that holds an array of photons. The reason the latter is not possible is that when a `Time` holds an array of times, all of them must have the same `location`.

sorry @paulray, which `location` are we talking about here? For ground-based observatories, the `EarthLocation` is fixed

**Comment by paulray** [comment]
True, for the special case where all TOAs are from a single Earth-fixed observatory, one could make a single `Time` object with an array of MJDs for a single `location`. But, PINT's TOA tables need to handle multiple observatories and moving (e.g. spacecraft) observatories.

**Comment by scottransom** [comment]
It might be possible to special case truly single-observatory, fixed location TOA table generation.  I'm not sure how much work that would be (or truly how much time it would save).  But it is definitely worth looking into.

**Comment by aarchiba** [comment]
This is a prime application for profiling! I suspect we'd find a lot of the time was spent in actual computations, in time system conversions and Earth ephemeris evaluations. It's actually pretty straightforward to use the python profiler on scripts, test cases, or our whole test suite, and it can be surprising where we are spending time. For example, during fitting the linear algebra uses a tiny fraction of the time, while evaluating the models and their derivatives takes most of it.

**Comment by paulray** [comment]
There are profiling suites and examples in PINT/profiling that @champagne-cmd wrote. 
You can do something like this:
```
run_profile.py bench_load_TOAs.py
```
Then modify that script to load a bunch of photon TOAs instead of the J0740 .tim.
You can view the output PDF, or use `cprofilev` to view it interactively.

**Comment by kerrm** [comment]
This could certainly be true, but it would mean astropy+erfa are much
slower than lower level routines a la tempo2.  I haven't timed it recently,
but I'm sure tempo2 is at least 10x faster when phasing up Fermi data.  It
would be good to know either way.

**Comment by matteobachetti** [comment]
One thing we do in Stingray when dealing with already barycentered data is creating a small-ish number of fake TOAs between the start and the end of the observation, calculate the correction on those, and interpolate the correction for the actual event times (often various orders of magnitude more numerous). In the timescales of a typical X-ray observation, it works decently well, and I was about to propose to use something like this in PINT as well, at least for data already referred to the TDB.

**Comment by paulray** [comment]
Agree, that can work well. What Matthew and I did for Fermi was to make TEMPO polycos then use those to compute phases for photons. Evaluating polycos is much faster than full model evaluations in PINT. We should probably add some code to PINT to make this easy for large photon datasets.

**Comment by rlopezcoto** [comment]
that sounds like a good solution. @matteobachetti do you have a plot of the difference between the interpolated event times correction and the real one calculated with their TOAs?
